lib/qos/qos.c

```diff
@@ -0,0 +1,336 @@
+
+	if (ctx.initialized) {
+		err = -EALREADY;
+		goto exit;
+	}
+
+	ctx.initialized = true;
+
+	LOG_DBG("Registering handler %p", evt_handler);
+	ctx.app_evt_handler = evt_handler;
+
+	/* Initializing linked list and delayed work. */
+	sys_slist_init(&ctx.pending_list);
+	k_work_init_delayable(&ctx.timeout_handler_work, timeout_handler_work_fn);
+
+exit:
+	k_mutex_unlock(&ctx_lock);
+	return err;
+}
+
+int qos_message_add(struct qos_data *message)
+{
+	int err = 0, ret;
+	struct qos_evt evt = {
+		.type = QOS_EVT_MESSAGE_NEW,
+		.message = *message
+	};
+
+	k_mutex_lock(&ctx_lock, K_FOREVER);
+
+	if (message == NULL) {
+		LOG_ERR("Message cannot be NULL");
+		err = -EINVAL;
+		goto exit;
+	}
+
+	/* Only ACK_REQUIRED messages are added to the internal list. */
+	if (qos_message_has_flag(message, QOS_FLAG_RELIABILITY_ACK_REQUIRED)) {
+		ret = list_append(message);
+		if (ret < 0) {
+			LOG_WRN("No list entries available, error: %d", ret);
+			evt.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST;
+			notify_event(&evt);
+			err = -ENOMEM;
+			goto exit;
+		}
+
+		/* Increment notified count before the callback. */
+		ctx.list_internal[ret].message.notified_count++;
+		evt.message = ctx.list_internal[ret].message;
+		notify_event(&evt);
+	} else {
+
+		/* If the message does not carry the ACK_REQUIRED flag, its notified as a new
+		 * message before it is removed from the library.
+		 */
+		notify_event(&evt);
+
+		evt.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST;
+		notify_event(&evt);
+	}
+
+	/* Start the internal timer if it is not running. */
+	if (!k_work_delayable_is_pending(&ctx.timeout_handler_work)) {
+		k_work_reschedule(&ctx.timeout_handler_work,
+				  K_SECONDS(CONFIG_QOS_MESSAGE_NOTIFY_TIMEOUT_SECONDS));
+	}
+
+exit:
+	k_mutex_unlock(&ctx_lock);
+	return err;
+}
+
+int qos_message_remove(uint32_t id)
+{
+	int err;
+
+	k_mutex_lock(&ctx_lock, K_FOREVER);
+
+	err = list_remove(id);
+	if (err) {
+		LOG_ERR("Failed removing message from list, message ID not found");
+		goto exit;
+	}
+
+	/* If the removed message is the last in the pending list, we stop the internal timer. */
+	if (sys_slist_is_empty(&ctx.pending_list)) {
+		LOG_DBG("QoS list is empty, cancel ongoing delayed work");
+		k_work_cancel_delayable(&ctx.timeout_handler_work);
+	}
+
+exit:
+	k_mutex_unlock(&ctx_lock);
+	return err;
+}
+
+void qos_message_print(const struct qos_data *message)
+{
+	LOG_DBG("Notified count: %d", message->notified_count);
+	LOG_DBG("Message heap_allocated: %d", message->heap_allocated);
+	LOG_DBG("Message ID: %d", message->id);
+	LOG_DBG("Message Buffer pointer: %p", message->data.buf);
+	LOG_DBG("Message Buffer length: %d", message->data.len);
+	LOG_DBG("Message Flags: %x", message->flags);
+	LOG_DBG("Message type: %d", message->type);
+}
+
+bool qos_message_has_flag(const struct qos_data *message, uint32_t flag)
+{
+	return ((message->flags & flag) == flag) ? true : false;
+}
+
+uint16_t qos_message_id_get_next(void)
+{
+	if (message_id_next == UINT16_MAX) {
```

Comment:
`message_id_next` can be a static variable in this function as it's only used here, or possibly in the `ctx` variable